drivers/docker/stats.go

```diff
@@ -141,16 +137,40 @@ func (h *taskHandle) collectStats(ctx context.Context, destCh *usageSender, inte
 			retry++
 			continue
 		}
+
+		var stats containerapi.Stats
+		statsStringScanner := bufio.NewScanner(statsReader.Body)
+
+		// StatsResponseReader that the SDK returns is somewhat unpredictable. Sometimes
+		// during 1 interval window, it will respond with multiple Stats objects,
+		// sometimes it won't. The reader won't close until the container stops, so it's
+		// up to us to digest this stream carefully.
+		// The scanner below gets just one line and sends it to the channel.
+		for statsStringScanner.Scan() {
+			if err := json.Unmarshal(statsStringScanner.Bytes(), &stats); err != nil {
+				h.logger.Error("error unmarshalling stats data for container", "error", err)
+				break
+			}
+			statsCh <- &stats
+			break
```

Comment:
Seeing this `break` makes me question the outer loop. With each pass through the loop, we're spawning a _new_ `statsCh` and goroutine that receives on that channel to send to the destination channel periodically. But it doesn't look like anyone ever closes that channel. And then we have this inner loop that's pulling of Docker's API but throwing out any backlog. Which means the logic in `dockerStatsCollector` to ensure we're sending only the latest values doesn't work anymore.

Reply:
_actually_, the more I think about this, the more I'm inclined to indeed remove the outer loop. We could implement interval timer inside the scanner loop, and do the backoff on the API call separately. Hold my coffee, I fix.

Reply:
have a look at [15ef385](https://github.com/hashicorp/nomad/pull/23966/commits/15ef385cf28c62d0e51625270d4b0bdf0ca9bc11)? Doesn't remove the outer loop (it's hard to do backoff retries without it), but improves the code a bit I think.